src/Geno/C++/GUI/Widgets/TextEdit.h

```diff
@@ -144,6 +149,76 @@ class TextEdit
 		None
 	};
 
+	struct LineSelectionItem
+	{
+		enum
+		{
+			None,
+			Selection,
+			Search,
+			SearchActive
+		};
+
+		Coordinate Start;
+		Coordinate End;
+
+		int Type;
+
+		LineSelectionItem( int Type = None )
+			: Type( Type )
+		{
+		}
+	};
+
+	struct SearchDialog
+	{
+		bool Searching     = false;
+		bool CaseSensitive = false;
+		bool WholeWord     = false;
+		int  ActiveItem    = -1;
+
+		std::string SearchTerm;
+
+		struct SearchResultGroups
+		{
+			// Results will be grouped in it's own vector for every 1000 lines
+			std::vector< std::vector< LineSelectionItem* > > Groups;
+			// A vector of all the results in order
+			std::vector< LineSelectionItem > Result;
+
+			SearchResultGroups();
+
+			std::vector< LineSelectionItem* >& GetGroup( int LineIndex );
+			bool                               GroupExist( int LineIndex );
+			void                               AddResult( const LineSelectionItem& Item );
+			void                               UpdateGroups();
+			void                               Clear();
+			size_t                             Size();
+			bool                               Empty();
+
+			LineSelectionItem& operator[]( int Index );
+		};
+
+		struct SearchInstance
+		{
+			enum
+			{
+				Running,
+				Stopping,
+				Stopped,
+				HasResult
+			};
+
+			std::thread         Thread;
+			std::string         SearchTerm;
+			SearchResultGroups* Result;
+			int                 State;
+		};
+
+		std::vector< SearchInstance* > SearchInstances;
+		SearchResultGroups*            SearchResult = nullptr;
+	};
```

Comment:
Looks like this should be refactored into its own class

Reply:
But it's only ever used in textedit. Makes most sense it lives in textedit IMO.

Reply:
While I agree, it also takes up a lot of real estate and makes the file harder to skim over.

Reply:
True. So do it get it's own file or just outside the textedit class?

Reply:
its own file, preferrably

Reply:
well that's gonna be a massive pain as it depends on thing in textedit

Reply:
All right, keep it this way for now. I'll have to figure out a way for us to split up files in the future.